src/Client/ConnectionEstablisher.cpp

```diff
@@ -72,6 +74,16 @@ void ConnectionEstablisher::run(ConnectionEstablisher::TryResult & result, std::
         ProfileEvents::increment(ProfileEvents::DistributedConnectionUsable);
         result.is_usable = true;
 
+        bool is_readonly = table_status_it->second.is_readonly;
+        if (insert && is_readonly)
```

Comment:
Passing `insert` flag here is too intrusive. Logically, connection is almost nothing to do with operations on top of it (it only provides connections with some properties). I'd suggest adding `read_only` flag into `TryResult`, so you can check in upper level code

Reply:
Make sense, fixed in c0e06cbc11269e7841ca3a0e863d5a1ff8cce6d6